treematcher.py

```diff
@@ -1,286 +1,178 @@
+    def get_cached_attr(self, attr_name, node, leaves_only=False):
+        """
+        easy access to cached attributes on trees.
 
-        try:
-            self.cache_flag = kargs.pop('cache')
+        :param attr_name: any attribute cached in tree nodes (e.g., species,
+         name, dist, support, etc.)
 
-        except KeyError:
-            self.cache_flag = False
+        :param node: The pattern tree containing the cache
 
+        :leaves_only: If True, return cached values only from leaves
 
-        if len(args) > 0:
-            pattern_string = args[0].strip()
+        :return: cached values for the requested attribute (e.g., Homo sapiens,
+         Human, 1.0, etc.)
 
-            if pattern_string.find('@') == -1:
-                is_exact = True
-            else:
-                #cached_searches = ['.leaves', ".contains_species", ".contains", ".size"]
+        """
+        print "USING CACHE"
+        cache = self.leaves_cache if leaves_only else self.all_node_cache
+        values = [getattr(n, attr_name, None) for n in cache[node]]
+        return values
 
-                shortcut_functions=["contains_species", "contains_duplication", "contains_speciation", "contains"]
+    def get_leaves(self, node):
+        return self.leaves_cache[node]
 
-                for function in shortcut_functions:
-                    if pattern_string.find(function + '(') != -1:
-                        self.shortcut_functions += [function]
+    def get_descendants(self, node):
+        return self.all_node_cache[node]
 
+class _FakeCache(object):
+    """TreePattern cache emulator."""
+    def __init__(self):
+        pass
 
-                #if pattern_string.find("contains_duplication")!= -1 or pattern_string.find("contains_speciation") != -1:
-                #    self.evol_events_flag = 1
+    def get_cached_attr(self, attr_name, node, leaves_only=False):
+        """ Helper function to mimic the behaviour of a cache, so you functions can
+        refer to a cache even when this has not been created, thus simplifying code
+        writing. """
+        if leaves_only:
+            iter_nodes = node.iter_leaves
+        else:
+            iter_nodes = node.iter_descendants
 
+        values = [getattr(n, attr_name, None) for n in iter_nodes()]
+        return values
 
-        #Tree.__init__(self, *args, **kargs)
-        super(TreePattern, self).__init__(*args, **kargs)
-        self.temp_leaf_cache = {}
-        # all of this only happens in the root node
-        if len(args)>0:
+    def get_leaves(self, node):
+        return node.get_leaves()
 
-            for n in self.traverse():
-                n._extra_functions = self._extra_functions
+    def get_descendants(self, node):
+        return node.get_descendants()
 
-                if n.name != "NoName":
-                    self._parse_constraint(n, is_exact)
-                else:
-                    n.constraint = None
 
+class PatternSyntax(object):
```

Comment:
This object should contain all keywords, shortcut and help functions used in pattern expressions. 
I already implemented a way where everything within this class becomes the local scope (formerly `custom_vars` in the old code) passed to the python eval function. 
I think this is nice and cleaner. All @AlishaMechtley's functions are now moved here